demos/device_defender_for_aws/metrics_collector/freertos_plus_tcp/metrics_collector.c

```diff
@@ -0,0 +1,272 @@
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ * https://aws.amazon.com/freertos
+ * https://www.FreeRTOS.org
+ */
+
+/**
+ * @file metrics_collector.c
+ *
+ * @brief Functions used by the defender demo to collect metrics on the
+ * device's open ports and sockets. FreeRTOS+TCP tcp_netstat utility
+ * is used to collect this metrics.
+ */
+
+/* Standard includes. */
+#include <stdio.h>
+#include <ctype.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdint.h>
+
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "FreeRTOS_IP.h"
+
+/* FreeRTOS+TCP tcp_netstat utility include. */
+#include "tcp_netstat.h"
+
+/* Demo config. */
+#include "defender_demo_config.h"
+
+/* Interface include. */
+#include "metrics_collector.h"
+/*-----------------------------------------------------------*/
+
+MetricsCollectorStatus_t GetNetworkStats( NetworkStats_t * pOutNetworkStats )
+{
+    MetricsCollectorStatus_t status = MetricsCollectorSuccess;
+
+    MetricsType_t metrics = { 0 };
+    BaseType_t metricsStatus = 0;
+
+    configASSERT( pOutNetworkStats != NULL );
+
+    /* Start with everything as zero. */
+    memset( pOutNetworkStats, 0, sizeof( NetworkStats_t ) );
+
+    /* Get metrics from FreeRTOS+TCP tcp_netstat utility. */
+    metricsStatus = vGetMetrics( &metrics );
+
+    if( metricsStatus != 0 )
+    {
+        LogError( ( "Failed to acquire metrics from FreeRTOS+TCP tcp_netstat utility. Status: %d.",
+                    ( int ) metricsStatus ) );
+        status = MetricsCollectorCollectionFailed;
+    }
+
+    /* Fill our response with values gotten from FreeRTOS+TCP. */
+    if( status == MetricsCollectorSuccess )
+    {
+        LogDebug( ( "Network stats read. Bytes received: %u, packets received: %u, "
+                    "bytes sent: %u, packets sent: %u.",
+                    ( unsigned int ) metrics.xInput.uxByteCount,
+                    ( unsigned int ) metrics.xInput.uxPacketCount,
+                    ( unsigned int ) metrics.XOutput.uxByteCount,
+                    ( unsigned int ) metrics.XOutput.uxPacketCount ) );
+
+        pOutNetworkStats->bytesReceived = metrics.xInput.uxByteCount;
+        pOutNetworkStats->packetsReceived = metrics.xInput.uxPacketCount;
+        pOutNetworkStats->bytesSent = metrics.XOutput.uxByteCount;
+        pOutNetworkStats->packetsSent = metrics.XOutput.uxPacketCount;
+    }
+
+    return status;
+}
+/*-----------------------------------------------------------*/
+
+MetricsCollectorStatus_t GetOpenTcpPorts( uint16_t * pOutTcpPortsArray,
+                                          uint32_t tcpPortsArrayLength,
+                                          uint32_t * pOutNumTcpOpenPorts )
+{
+    MetricsCollectorStatus_t status = MetricsCollectorSuccess;
```

Comment:
Same comment as `GetNetworkStats()`, this function will be simpler, shorter, and more durable during maintenance if you initialize status to failure here, e.g:

```
     MetricsCollectorStatus_t status = MetricsCollectorFailure; 
```